File: webapp/src/main/scala/org/linnando/mixemulator/webapp/EditorComponent.scala
package org.linnando.mixemulator.webapp

import angulate2.core.AfterViewInit
import angulate2.router.Router
import angulate2.std._
import com.scalawarrior.scalajs.ace._
import org.scalajs.dom
import org.scalajs.dom.raw.{Blob, BlobPropertyBag, FileReader, URL}
import org.scalajs.dom.{File, FileList, UIEvent, html}

import scala.concurrent.ExecutionContext.Implicits.global
import scala.scalajs.js
import scala.util.{Failure, Success}

@Component(
  selector = "editor",
  templateUrl = "webapp/src/main/resources/editor.component.html",
  styleUrls = @@@("webapp/src/main/resources/editor.component.css")
)
class EditorComponent(router: Router, virtualMachineService: VirtualMachineService) extends AfterViewInit {
  var editor: Editor = _

  override def ngAfterViewInit(): Unit = {
    editor = ace.edit("editor")
    editor.setTheme("ace/theme/chrome")
    editor.setValue(virtualMachineService.text)
    editor.getSession().setMode("ace/mode/mixal")
    editor.getSession().setUseSoftTabs(true)
    editor.getSession().on("change", (_: js.Any) => virtualMachineService.text = editor.getValue())
  }

  def startFileUpload(): Unit = {
    dom.document.getElementById("inputFile").asInstanceOf[html.Input].click()
  }

  def uploadFile(files: FileList): Unit =
    if (files.length > 0) {
      val file = files(0)
      val reader = new FileReader()
      reader.onload = (event: UIEvent) => {
        val target = event.target.asInstanceOf[js.Dynamic]
        val programText = target.result.asInstanceOf[String].replaceAll("\t", "    ")
        editor.setValue(programText)
      }
      reader.readAsText(file)
    }

  def downloadFile(): Unit = {
    val blobParts = js.Array[String](virtualMachineService.text).asInstanceOf[js.Array[js.Any]]
    val options = js.Dynamic.literal("type" -> "text/plain").asInstanceOf[BlobPropertyBag]
    val blob = new Blob(blobParts, options)
    val url = URL.createObjectURL(blob)
    js.Dynamic.global.window.open(url)
  }

  def mode: String = virtualMachineService.mode

  def mode_=(value: String): Unit = virtualMachineService.mode = value

  def tracking: Boolean = virtualMachineService.tracking

  def tracking_=(value: Boolean): Unit = virtualMachineService.tracking = value

  def assemble(): Unit = virtualMachineService.assemble() onComplete {
    case Success(_) => router.navigate(js.Array("/vm"))
    case Failure(e) => ErrorPopup.show(e)
  }
}

File: webapp/src/app/editor/editor.component.ts
import { AfterViewInit, Component } from '@angular/core';
import { Router } from '@angular/router';
import * as ace from 'ace-builds';
import 'ace-builds/webpack-resolver';
import { BsModalService } from 'ngx-bootstrap/modal';
import { ErrorPopupComponent } from '../error-popup/error-popup.component';
import { VirtualMachineService } from '../virtual-machine.service';

@Component({
  selector: 'mix-editor',
  templateUrl: './editor.component.html',
  styleUrls: ['./editor.component.scss']
})
export class EditorComponent implements AfterViewInit {
  editor?: ace.Ace.Editor;

  constructor(private modalService: BsModalService,
    private router: Router,
    private virtualMachineService: VirtualMachineService) { }

  ngAfterViewInit(): void {
    this.editor = ace.edit('editor');
    this.editor.setTheme('ace/theme/chrome');
    this.editor.setValue(this.virtualMachineService.text);
    this.editor.getSession().setMode('ace/mode/mixal');
    this.editor.getSession().setUseSoftTabs(true);
    this.editor.getSession().on('change', (): void => {
      this.virtualMachineService.text = this.editor?.getValue() ?? '';
    });
  }

  startFileUpload(): void {
    document.getElementById('inputFile')?.click();
  }

  uploadFile(target: EventTarget | null): void {
    if (!(target instanceof HTMLInputElement)) {
      return;
    }
    const files = target.files;
    if (!files || files.length == 0) {
      return;
    }
    const file = files[0];
    const reader = new FileReader();
    reader.onload = (event: ProgressEvent<FileReader>) => {
      const fileData = event.target?.result;
      if (!(typeof fileData === 'string')) {
        return;
      }
      const programText = fileData.replace(/\t/g, '    ');
      this.editor?.setValue(programText);
    };
    reader.readAsText(file);
  }

  downloadFile(): void {
    const blob = new Blob([this.virtualMachineService.text], { 'type': 'text/plain' });
    const url = URL.createObjectURL(blob);
    window.open(url);
  }

  get mode(): string {
    return this.virtualMachineService.mode;
  }

  set mode(value: string) {
    this.virtualMachineService.mode = value;
  }

  get tracking(): boolean {
    return this.virtualMachineService.tracking;
  }

  set tracking(value: boolean) {
    this.virtualMachineService.tracking = value;
  }

  assemble(): void {
    try {
      this.virtualMachineService.assemble();
      this.router.navigate(['/vm']);
    } catch (e) {
      this.modalService.show(ErrorPopupComponent, { initialState: { exception: e } });
    }
  }
}
